Improve Large MemoryPool Performance - Sort + intelligent TX reverification

jsolman opened the pull request:
Update: Fairly extensive testing has been performed on this PR and it is now considered stable and ready for merging.

***Background***
When a large amount of transactions were broadcast to the chain, the overhead in the current design would cause performance issues. Also, new transactions with high fees, could be delayed. In the existing code base, after each block, transactions were removed and re-sent to the `Blockchain` mailbox in the order by highest fee. This could cause mailbox transaction messages to pile up. Also, since the `Blockchain` mailbox treated these transaction messages all at the same priority level, so new transactions added with a high fee would still need to wait for those ahead of it in the mailbox to be re-verified before getting to them.

***Root Cause***
A very large MemoryPool would lead to way too many transactions potentially needing to be re-verified each block. A lot of the re-verification is wasted effort also since the consensus will only take a subset (500) of such transactions per block. The ideal scenario would only be to reverify those that can make it into the next block.

***Summary of Change***
This change only re-verifies as many transactions as necessary to start off with a pool containing as many transactions as can be accepted in a block. To facilitate this, it is best to keep the transactions sorted so it can easily iterate the top transactions to be re-verified after persisting a block. Since it keeps transactions now sorted based on their priority by fee per byte, then fee, then TX hash, it can easily verify the transactions with minimal required effort. It also only attempts to re-verify transactions if it is at the header height. Additionally, it limits the time it will spend verifying transactions to add them back to the next block.

The mempool `Count` now includes all transactions that had been verified at some point that haven't become known to be invalid. The `VerifiedCount` and `UnverifiedCount` are also now exposed.

Closes #478 
Closes #427 
Closes #353 
Closes #358 
Closes #329 
Closes #443 
Closes #539
Closes #408 - May be debatable whether this should close this without a third category (aside from high and low priority) allowing a few completely free transactions to be persisted each block; but since we should not really promote hash war, I think it closes it.
Closes #428

vncoelho commented:
This will be surely a great change, Jeff.

erikzhang commented:
Maybe we should create a background thread to specifically validate transactions instead of porting the policy plugin.

jsolman commented:
I thought about having another thread for verifying transactions, but it may require some locking and add some additional complexity. I’ll think about it some more while I do some testing on this.

This didn’t really port the whole policy plugin, but it does need access to how many transactions can be verified per block and how many of those are low priority. It still can end up with more transactions verified than the policy accepts by the time the consensus happens though, so the policy is still necessary with this as the changes currently stand.

jsolman commented:
Whether I add an additional thread for verifying transactions or not, I do think it make sense to verify more transactions after `PersistCompleted` in the case it runs out of time before it could add the desired amount. On a powerful machine it should always finish probably, but since there will still be 8+ seconds (with 15 sec blocks) or so before the next consensus it makes sense to verify any more that can fit in with the remaining time available, and is a good case for a separate thread to be able push max allowed TX per block to the max. I will make an adjustment in some way to handle that.

Options:
1. send a new VerifyTXBatch message back to self that upon completion sends another VerifyTXBatch message while verified transactions are below the desired amount. This gives a chance for new transactions to come in between the batches.
2. use a background thread for verifying messages; it has to pause to allow verifying a new TX that comes in so that it can respond with relay reason in a timely fashion
3. Switch the Blockchain mailbox to have a real priority queue that for transactions considers their fees in order of message delivery — I don’t really like this because it’s not a good idea to leave a message hanging without a response to whether the transaction was relayed and why.

Option 1 is probably the easiest to implement and I may go that route for this PR. There aren’t that many other messages the Blockchain actor is processing while waiting for a new block and transaction verification is taking place, so there doesn’t seem like there would be a significant benefit for another thread for the verification purpose, but let me think about it some more.

erikzhang commented:
Option 1 is fine, we can send an `Idle` message to the actor and then handle some things.

erikzhang commented:
Send `Idle` message to actors when they are idle: #502

jsolman commented:
Thanks for the quick addition of the Idle message. I’ll update it to use the Idle message when it is working hours for me again.

shargon commented on neo/Settings.cs:
```diff
@@ -33,6 +35,8 @@ public Settings(IConfigurationSection section)
             this.SecondsPerBlock = GetValueOrDefault(section.GetSection("SecondsPerBlock"), 15u, p => uint.Parse(p));
             this.LowPriorityThreshold = GetValueOrDefault(section.GetSection("LowPriorityThreshold"), Fixed8.FromDecimal(0.001m), p => Fixed8.Parse(p));
+            this.MaxTransactionsPerBlock = GetValueOrDefault(section.GetSection("MaxTransactionsPerBlock"), (uint) 500, uint.Parse);
+            this.MaxFreeTransactionsPerBlock = GetValueOrDefault(section.GetSection("MaxTransactionsPerBlock"), (uint) 20, uint.Parse);
```
MaxFreeTransactionsPerBlock in settings

jsolman replied:
nice catch, thank you

shargon commented on neo/Ledger/MemoryPool.cs:
```diff
@@ -4,161 +4,312 @@
+            
+            public int CompareTo(object obj)
+            {
+                if (obj == null) return 1;
```
if (!(obj is PoolItem otherItem))

jsolman replied:
Added that check

jsolman commented:
@shargon thanks for that fix

erikzhang commented:
Better `Idle` message: #505

igormcoelho commented on neo/Ledger/MemoryPool.cs:
```diff
@@ -4,161 +4,357 @@
+        
+        // These two are not expected to be hit, they are just safegaurds. 
+        private static readonly double MaxSecondsToReverifyHighPrioTxPerIdle = (double) Blockchain.SecondsPerBlock / 15;
+        private static readonly double MaxSecondsToReverifyLowPrioTxPerIdle = (double) Blockchain.SecondsPerBlock / 30;
```
15/3 = 5  15/5=3
15/15=1  15/30=0.5
safeguards are intended to be smaller?

jsolman replied:
Yes the safeguards only validate like 10 messages at a time; should never take even close to a second or half a second; they are only used when validating messages while Blockchain is idle. They just ensure the blockchain actor is never stuck busy for a long time.

Note on this PR: I still have an additional change coming shortly to make the Idle message re-queue itself while there are still more unconfirmed transactions.

erikzhang commented on neo/Ledger/MemoryPool.cs:
```diff
@@ -4,162 +4,360 @@
-        static void RemoveLowestFee(ConcurrentDictionary<UInt256, PoolItem> pool, int count)
+        public IEnumerable<Transaction> GetVerifiedTransactions()
+        {
+            IEnumerator verifiedTxEnumerator = _unsortedTransactions.Select(p => p.Value.Transaction).GetEnumerator();
```
```c#
return _unsortedTransactions.Select(p => p.Value.Transaction);
```

jsolman replied:
Thanks, I’ll simplify it to that.

erikzhang commented on neo/Ledger/MemoryPool.cs:
```diff
@@ -4,162 +4,360 @@
     internal class MemoryPool : IReadOnlyCollection<Transaction>
     {
-        private class PoolItem
+        private class PoolItem : IComparable
```
`IComparable<PoolItem>`

jsolman replied:
Changed.

erikzhang commented:
What about reordering transactions each time? Our memory pool can only contain up to 50,000 transactions, and reordering takes only 44 milliseconds.

```c#
void TestSort()
{
    Random rand = new Random();
    byte[] buffer = new byte[sizeof(long)];
    long[] values = new long[50000];
    for (int i = 0; i < values.Length; i++)
    {
        rand.NextBytes(buffer);
        values[i] = BitConverter.ToInt64(buffer, 0);
    }
    Stopwatch stopwatch = Stopwatch.StartNew();
    values = values.OrderBy(p => p).ToArray();
    stopwatch.Stop();
    Console.WriteLine($"{stopwatch.ElapsedMilliseconds} ms"); // 44 ms
}
```

jsolman commented:
Sorting on persist may be faster; I'll look tomorrow.

jsolman commented:
Actually when the pool is full; sorting each time to see what to throw away is definitely going to be slower than this code.

jsolman commented:
If anyone has time to do some testing on this and report results here I would appreciate it. I want to do more testing myself with injecting high tx throughput in a private net, but I have other priorities right now. Thank you.

erikzhang commented:
I did more tests:

```c#
class PoolItem : IComparable<PoolItem>
{
    public object Transaction;
    public DateTime Timestamp;
    public long FeePerByte;

    public int CompareTo(PoolItem other) => FeePerByte.CompareTo(other.FeePerByte);
}

PoolItem[] GetItems(int count)
{
    Random rand = new Random();
    byte[] buffer = new byte[sizeof(long)];
    PoolItem[] items = new PoolItem[count];
    for (int i = 0; i < items.Length; i++)
    {
        rand.NextBytes(buffer);
        items[i] = new PoolItem
        {
            Transaction = new object(),
            Timestamp = DateTime.Now,
            FeePerByte = BitConverter.ToInt64(buffer, 0)
        };
    }
    return items;
}

void TestSort()
{
    PoolItem[] items = GetItems(50000);
    Stopwatch stopwatch = Stopwatch.StartNew();
    items = items.OrderBy(p => p).ToArray();
    stopwatch.Stop();
    Console.WriteLine($"{stopwatch.ElapsedMilliseconds}ms"); // 47 ms
}
```

Sorting 50,000 PoolItems takes only 47 milliseconds.

jsolman commented:
@erikzhang Ok, but what are you suggesting to changed based on your findings exactly? Did you compare that to the time it takes to add a single transaction and then remove a single transaction in the sorted set when it is already sorted with 50,000 items in it? Because that has to be much faster than the 47 ms for sorting it each time.

The code really wants them to be sorted each time a transaction comes in to know which old one to discard. At the best case if Linq is doing the optimal thing to remove the lowest fee transaction it is an O(n) operation to find that transaction with a non sorted structure as opposed to an O(log(n)) operation with the sorted set.

Are you suggesting that the only time the code needs the transactions sorted is after persisting a block? I think for the verified transactions it is best if they are always sorted to make it fast to remove the old transactions. Although since they are added and removed one at a time as transactions come in, a min heap would actually be faster for the verified transactions then the sorted set since it is faster for removing just the smallest item and reheapifying.

For the data structure for the unverified transactions, I agree that it could just add those to a list and then sort that list if that is what you are suggesting by adding the sort profile code. could add them to a list and then sort that list. Is that what you are suggesting I should change?

jsolman commented:
The two improvements that could still be made to this code as I see it:

1. I could switch the verified high and low priority sorted data structures from sorted set to min heap (but I may need to add a dependency to the code base for that data structure)
2. I could switch the unverified sorted data structure from a sorted set to a list and sort the list instead since it only has to be sorted once after persist and then can just always remove from the end when reverifying.

Those are the two things I can think of to improve the performance a bit.

Additionally, I could expose a method that gets the verified transactions in sorted order so the consensus could get them already sorted instead of using the concurrent dictionary that is inspected, but it would need to do some locking.

jsolman commented:
Optimization number 2 actually means also that after persisting a block a merge sort algorithm would need to be employed to add the verified transactions that didn’t make it into a block to the unverified transactions to optimize that operation. 

Using the sorted set for unverified transactions avoids that complexity.

Really if the time related to this sorted set is small compared to the time actually verifying the transactions our discussion about sorting vs inserting to a sorted structure is moot and might as well go with this implementation that is fairly simple and should be resonably good performance.

erikzhang commented:
I have no problem with the current solution.
But I recommend that in order to ensure thread safety, `MemoryPool` only provides `GetUnverifiedTransactions(int count)` and `GetVerifiedTransactions()`, and then handles all transaction validation in `Blockchain`.

vncoelho commented on neo/Ledger/MemoryPool.cs:
```diff
@@ -4,162 +4,358 @@
+        private readonly SortedSet<PoolItem> _unverifiedSortedLowPriorityTransactions = new SortedSet<PoolItem>();
+        
+        private int MaxHighPriorityTxsPerBlock => Settings.Default.MaxTransactionsPerBlock 
+            - Settings.Default.MaxFreeTransactionsPerBlock;
```
I think that we can have `Settings.Default.MaxTransactionsPerBlock` transactions in a block if there are no free transactions.

jsolman replied:
I could rename it since it isn't really the max, it's just how many it takes after a block is verified. More unverified transactions will now be pulled in anyway when processing the Idle message, and it is the plugin code that decides how many of the verified transactions to take currently. Also, at the time that the transactions are re-verified after a block is persisted it is unknown how many new high priority transactions will come in during the time until the next block.

jsolman replied:
I addressed this comment and just use Settings.Default.MaxTransactionsPerBlock directly for high priority transactions.

vncoelho commented on neo/Ledger/MemoryPool.cs:
```diff
@@ -4,162 +4,358 @@
+            return ret;
+        }
+        
+        // Note: this isn't used in Fill during consensus, fill uses GetVerifiedTransactions()
```
Fill should get the Sorted, right? Function GetVerifiedTransactions is returning unsorted.

jsolman replied:
Yeah, it needs locking to return the transactions sorted. I expected to make that change before this PR is accepted and merged. I'll probably add a R/W lock for it.

jsolman replied:
Made changes. It gets them sorted now for fill.

vncoelho replied:
You are right, @erikzhang https://github.com/neo-project/neo-plugins/blob/db04e756a64c149e86d117fe89879c72f319c7df/SimplePolicy/SimplePolicyPlugin.cs#L58

But sort the sorted should be almost O(n), yes? @igormcoelho 
Thus, better to send it sorted anyway.

vncoelho commented on neo/Ledger/MemoryPool.cs:
```diff
@@ -4,162 +4,358 @@
+            // NOTE: This really shouldn't be necessary any more and can actually be counter-productive, since
+            //       they can be re-added anyway and would incur a potential addtional verification; setting to 1000
+            //       blocks for now. If transactions want to only be valid for short times they can include a script.
+            var lowPriorityCutOffTime = DateTime.UtcNow.AddSeconds(-Blockchain.SecondsPerBlock * 1000);
```
In this sense, maybe this cutoff can be removed.

jsolman replied:
We could remove it, but I guess if there were consistently low priority transactions transactions with small fees for a long period of time (like say 24 hours) it would be desirable to eventually expire off the completely free transactions with no fees to allow them to be sent again with a fee. Using 1000 blocks here achieves that; might as well keep it here; though clients should still use a best practice of attaching a script to expire transactions sooner; this is more like a maximum for them to use. Of course it is possible they could be expired sooner if the pool is filled up with 50000 higher fee transactions.

vncoelho commented on neo/Ledger/MemoryPool.cs:
```diff
@@ -4,162 +4,358 @@
+
+            // If we know about headers of future blocks, no point in verifying transactions from the unverified tx pool
+            // until we get caught up.
+            if (block.Index < Blockchain.Singleton.HeaderHeight)
```
This is a good think and will surely help Chain Sync! Nice catch, Jeff.

vncoelho commented on neo/Ledger/MemoryPool.cs:
```diff
@@ -4,162 +4,358 @@
 
-        public bool TryGetValue(UInt256 hash, out Transaction tx)
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private int ReverifyLowPriorityTransactions(int count, double secondsTimeout, Snapshot snapshot)
```
This method and ReverifyHighPriorityTransactions and very similar.

Then, I thought if this Objects could be created as pairs, the logic would become more compact and, I believe, more simple. What do you think?

jsolman replied:
@vneolho well based on the methods exposed but sorted set they have to use different structure because you need to take 20 from low priority and 480 from high priority. If the sorted set exposed iterating from a specific item (which the underlying binary tree supports) it would work, but unfortunately SortedSet designers didn’t expose that ability.

vncoelho replied:
A pair can have different sizes. .

I mean, an array with 2 objects, one for high and another for low.
Then, you just pass the index to the function.

jsolman replied:
Ok, in that case, you just meant to remove the duplicated logic. Sure; I'll take a few minutes to adjust to remove the duplicated logic. It will need to pass a few other parameters though, and it can just pass the set and not need to add a Pair.

jsolman replied:
Done.

vncoelho commented on neo/Ledger/MemoryPool.cs:
```diff
@@ -4,162 +4,358 @@
+                maxToVerify -= ReverifyHighPriorityTransactions(verifyCount, MaxSecondsToReverifyHighPrioTxPerIdle, 
+                    snapshot);
+                
+                if (maxToVerify == 0) maxToVerify++;
```
Is this for ensuring  the verification of, at least, 1 Low Priority tx?

jsolman replied:
Yes.

jsolman commented on neo/Ledger/MemoryPool.cs:
```diff
@@ -4,162 +4,376 @@
+                if (block.Index < Blockchain.Singleton.HeaderHeight)
+                    return;
+
+                ReverifyTransactions(_sortedHighPrioTransactions, _unverifiedSortedHighPriorityTransactions,
```
If the node is being bombarded with new transactions constantly that are low priority it will prevent Idle from running much and re-verifying the highest fee transactions. In order to ensure high fee transactions get in a block first it is important to immediately give some time to reverify them right after the block persists.

vncoelho commented:
I like the new PR name, @jsolman.

erikzhang commented:
I wrote a `SortedKeyedSet` class that might be useful.

```c#
/// <summary>
/// Provides the abstract base class for a sorted set whose keys are embedded in the values.
/// </summary>
/// <typeparam name="TKey">The type of keys in the set.</typeparam>
/// <typeparam name="TItem">The type of items in the set.</typeparam>
public abstract class SortedKeyedSet<TKey, TItem> : ISet<TItem>
    where TItem : IComparable<TItem>
{
    public readonly object SyncRoot = new object();
    private readonly Dictionary<TKey, TItem> dictionary = new Dictionary<TKey, TItem>();
    private readonly SortedSet<TItem> sorted = new SortedSet<TItem>();

    public TItem this[TKey key]
    {
        get
        {
            lock (SyncRoot)
            {
                return dictionary[key];
            }
        }
    }

    public int Count
    {
        get
        {
            lock (SyncRoot)
            {
                return dictionary.Count;
            }
        }
    }

    public bool IsReadOnly => false;

    public TItem Max
    {
        get
        {
            lock (SyncRoot)
            {
                return sorted.Max;
            }
        }
    }

    public TItem Min
    {
        get
        {
            lock (SyncRoot)
            {
                return sorted.Min;
            }
        }
    }

    public bool Add(TItem item)
    {
        TKey key = GetKeyForItem(item);
        lock (SyncRoot)
        {
            if (dictionary.ContainsKey(key))
                return false;
            dictionary.Add(key, item);
            sorted.Add(item);
            return true;
        }
    }

    void ICollection<TItem>.Add(TItem item)
    {
        Add(item);
    }

    public void Clear()
    {
        lock (SyncRoot)
        {
            dictionary.Clear();
            sorted.Clear();
        }
    }

    public bool Contains(TItem item)
    {
        return ContainsKey(GetKeyForItem(item));
    }

    public bool ContainsKey(TKey key)
    {
        lock (SyncRoot)
        {
            return dictionary.ContainsKey(key);
        }
    }

    public void CopyTo(TItem[] array, int arrayIndex)
    {
        lock (SyncRoot)
        {
            foreach (TItem item in sorted)
                array[arrayIndex++] = item;
        }
    }

    public void ExceptWith(IEnumerable<TItem> other)
    {
        lock (SyncRoot)
        {
            foreach (TItem item in other)
            {
                TKey key = GetKeyForItem(item);
                if (dictionary.Remove(key))
                    sorted.Remove(item);
            }
        }
    }

    public IEnumerator<TItem> GetEnumerator()
    {
        lock (SyncRoot)
        {
            foreach (TItem item in sorted)
                yield return item;
        }
    }

    IEnumerator IEnumerable.GetEnumerator()
    {
        return GetEnumerator();
    }

    protected abstract TKey GetKeyForItem(TItem item);

    public TItem GetOrAdd(TKey key, TItem item)
    {
        lock (SyncRoot)
        {
            if (dictionary.TryGetValue(key, out TItem result))
                return result;
            dictionary.Add(key, item);
            sorted.Add(item);
            return item;
        }
    }

    public void IntersectWith(IEnumerable<TItem> other)
    {
        ISet<TItem> set = other as ISet<TItem> ?? new HashSet<TItem>(other);
        lock (SyncRoot)
        {
            foreach (var pair in dictionary.ToArray())
                if (!set.Contains(pair.Value))
                {
                    dictionary.Remove(pair.Key);
                    sorted.Remove(pair.Value);
                }
        }
    }

    public bool IsProperSubsetOf(IEnumerable<TItem> other)
    {
        ISet<TItem> set = other as ISet<TItem> ?? new HashSet<TItem>(other);
        lock (SyncRoot)
        {
            if (sorted.Count >= set.Count)
                return false;
            foreach (TItem item in sorted)
                if (!set.Contains(item))
                    return false;
            return true;
        }
    }

    public bool IsProperSupersetOf(IEnumerable<TItem> other)
    {
        lock (SyncRoot)
        {
            int n = 0;
            foreach (TItem item in other)
            {
                n++;
                TKey key = GetKeyForItem(item);
                if (!dictionary.ContainsKey(key))
                    return false;
            }
            return dictionary.Count > n;
        }
    }

    public bool IsSubsetOf(IEnumerable<TItem> other)
    {
        ISet<TItem> set = other as ISet<TItem> ?? new HashSet<TItem>(other);
        lock (SyncRoot)
        {
            foreach (TItem item in sorted)
                if (!set.Contains(item))
                    return false;
            return true;
        }
    }

    public bool IsSupersetOf(IEnumerable<TItem> other)
    {
        lock (SyncRoot)
        {
            foreach (TItem item in other)
            {
                TKey key = GetKeyForItem(item);
                if (!dictionary.ContainsKey(key))
                    return false;
            }
            return true;
        }
    }

    public bool Overlaps(IEnumerable<TItem> other)
    {
        lock (SyncRoot)
        {
            foreach (TItem item in other)
            {
                TKey key = GetKeyForItem(item);
                if (dictionary.ContainsKey(key))
                    return true;
            }
            return false;
        }
    }

    public bool Remove(TKey key)
    {
        lock (SyncRoot)
        {
            if (!dictionary.TryGetValue(key, out TItem item))
                return false;
            dictionary.Remove(key);
            sorted.Remove(item);
            return true;
        }
    }

    public bool Remove(TItem item)
    {
        TKey key = GetKeyForItem(item);
        lock (SyncRoot)
        {
            if (!dictionary.Remove(key)) return false;
            sorted.Remove(item);
            return true;
        }
    }

    public bool SetEquals(IEnumerable<TItem> other)
    {
        ISet<TItem> set = other as ISet<TItem> ?? new HashSet<TItem>(other);
        lock (SyncRoot)
        {
            if (dictionary.Count != set.Count)
                return false;
            foreach (TItem item in set)
            {
                TKey key = GetKeyForItem(item);
                if (!dictionary.ContainsKey(key))
                    return false;
            }
            return true;
        }
    }

    public void SymmetricExceptWith(IEnumerable<TItem> other)
    {
        lock (SyncRoot)
        {
            foreach (TItem item in other)
            {
                TKey key = GetKeyForItem(item);
                if (dictionary.Remove(key))
                {
                    sorted.Remove(item);
                }
                else
                {
                    dictionary.Add(key, item);
                    sorted.Add(item);
                }
            }
        }
    }

    public bool TryGetValue(TKey key, out TItem item)
    {
        lock (SyncRoot)
        {
            return dictionary.TryGetValue(key, out item);
        }
    }

    public bool TryRemove(TKey key, out TItem item)
    {
        lock (SyncRoot)
        {
            if (!dictionary.TryGetValue(key, out item))
                return false;
            dictionary.Remove(key);
            sorted.Remove(item);
            return true;
        }
    }

    public void UnionWith(IEnumerable<TItem> other)
    {
        lock (SyncRoot)
        {
            foreach (TItem item in other)
            {
                TKey key = GetKeyForItem(item);
                if (!dictionary.ContainsKey(key))
                {
                    dictionary.Add(key, item);
                    sorted.Add(item);
                }
            }
        }
    }
}
```

erikzhang commented:
Updated it with `ReaderWriterLockSlim`. @jsolman

jsolman commented:
@erikzhang For methods like GetEnumerator where the read lock will be held during iteration, perhaps it is better to lock, copy, unlock, then iterate the copy so that if the code that is iterating blocks for some reason it doesn’t leave the lock held.

erikzhang commented:
I am worried that if you do not lock these collections carefully, as more people join the development, there will be bugs in this place that will be difficult to debug in the future.

erikzhang commented:
Modified again. "lock, copy, unlock, then iterate the copy"

jsolman commented:
Well the advantage of switching would be that the Count will be more accurate in the case persist is in progress and iterating the mempool will never potentially give the same transaction twice at the cost of some performance.

`GetSortedVerifiedTransactions`  is always consistent though.

jsolman commented:
> I have no problem with the current solution.
But I recommend that in order to ensure thread safety, MemoryPool only provides GetUnverifiedTransactions(int count) and GetVerifiedTransactions(), and then handles all transaction validation in Blockchain.

@erikzhang does that mean that you would like the RpcServer to be reporting the Union of both verified and unverified though or just verified?
Should the MemPool have a new method that returns both verified and unverified items atomically in the same call and should the RpcServer also have a method that provides both?

erikzhang commented:
RpcServer needs to be able to distinguish between verified and unverified transactions.

jsolman commented:
How about if I add optional parameters on the RpcServer call for `getrawmempool` to get verified and unverified transactions back as a json object containing both `verified` and `unverified` jobject arrays? I will also make it include the block index for which the verified transactions are verified.

I’ll adjust this PR so the Default for getrawmempool will remain as the current behavior of only giving back verified transactions at whatever the node’s current block height happens to be. 

Note that if the node is behind header height with these changes it won’t verify any transactions and getrawmempool will return an empty array then by default.

igormcoelho commented:
I think Transaction class should have a Fixed8 Ratio => Transaction.NetworkFee / Transaction.Size
This is useful is many places. What do you think ?

apheliontoken commented on neo/Network/P2P/Payloads/Transaction.cs:
```diff
@@ -97,6 +97,8 @@ public virtual Fixed8 NetworkFee
 
         public virtual Fixed8 SystemFee => Settings.Default.SystemFee.TryGetValue(Type, out Fixed8 fee) ? fee : Fixed8.Zero;
 
+        public readonly Fixed8 FeePerByte => NetworkFee / Size;
```
@igormcoelho readonly isn't valid here; so this change you made doesn't build.
Also, I think this would end up doing the division every time the value is accessed, which is sub-optimal.

jsolman replied:
@igormcoelho readonly isn't valid here; so this change you made doesn't build.
Also, I think this would end up doing the division every time the value is accessed, which is sub-optimal.

igormcoelho commented:
Oh, sorry, my bad. Do you think we can cache this value on transaction? Its already being recomputed on neo-plugins, perhaps we could do something thats efficient and reused everywhere.

jsolman commented:
@igormcoelho Yes, there should be a way to do something that has it computed only when necessary. Feel free to make those changes if you have time now and then adjust this again if you want. I left `FeePerByte` in transaction, but fixed it so it would compile. Also, I restored caching it in `PoolItem` for now.

igormcoelho commented on neo/Network/P2P/Payloads/Transaction.cs:
```diff
@@ -97,7 +97,17 @@ public virtual Fixed8 NetworkFee
+                    _feePerByte = NetworkFee / Size;
+                return _feePerByte;
+            }
+        }
```
I've seen things like this in many classes... it's nice, a typical C# thing, never seen on C++ before. Guess it should work :)

igormcoelho commented:
Ok, decided to open another PR just for this... so, if it's accepted before this one, we can start using in plugins and other parts too :)

igormcoelho commented on neo/Ledger/MemoryPool.cs:
```diff
@@ -4,162 +4,378 @@
             {
                 Transaction = tx;
                 Timestamp = DateTime.UtcNow;
+                FeePerByte = Transaction.FeePerByte;
```
I guess this line needs to go out again... do you think it's better to revert all my past commits? Sorry for the trouble man, I was just trying to help hahahah

jsolman replied:
no problem

jsolman commented:
Thanks, can adjust once #516 gets accepted and merged

jsolman commented:
@erikzhang I took a look at using `class TxPool : SortedKeyedSet<UInt256, PoolItem>`, by using 4 `TxPool`, high and low priority verified and high and low priority unverified. However, I think it is more efficient to use the data structures directly in MemoryPool for less locking overhead. Also, it is advantageous to use a single read/write lock to maintain consistency of the two pools (often it may be removing from unverified and adding to verified) and if a read came in the middle to get both it wouldn’t guarantee a consistent state.

jsolman commented:
Behavior should no longer be different for `RpcServer` and `ProtocolHandler` in terms of returning verified transactions and I added support to RpcServer for passing a flag to `getrawmempool` to return verified, unverified hashes as well as the block height.

jsolman commented:
> I am worried that if you do not lock these collections carefully, as more people join the development, there will be bugs in this place that will be difficult to debug in the future.

The `MemoryPool` is now only exposed through public methods in Blockchain. They should all be safe now. Have a look and let me know. Note that there are some places where order doesn't matter and it is faster to get the verified transactions unsorted (iterating the Dictionary is faster than the SortedSet).

erikzhang commented on neo/Network/RPC/RpcServer.cs:
```diff
@@ -297,7 +297,21 @@ private JObject Process(string method, JArray _params)
+                        Blockchain.Singleton.GetMemoryPoolSeparateVerfiedAndUnverified(
+                            out IEnumerable<Transaction> verifiedTransactions,
+                            out IEnumerable<Transaction> unverifiedTransactions);
+                        json["height"] = Blockchain.Singleton.Height;
```
Although `Blockchain.Singleton.Height` is called immediately after `GetMemoryPoolSeparateVerfiedAndUnverified()`, consistency is not guaranteed. It is recommended to get the block height first, then call `GetMemoryPoolSeparateVerfiedAndUnverified()`.

jsolman replied:
I’ll switch to reading the height before as you suggest.

jsolman replied:
Reading it before still doesn’t guarantee consistency unfortunately. Perhaps it Would better if the MemoryPool also kept the height for which it has last verified a transaction that it also protects with the write lock. Then it can also output that from `GetMemoryPoolSeparateVerifiedAndUnverified`.

erikzhang replied:
Transactions that are verified at lower height may not be verifiable at higher height; however, transactions that are verified at higher height are safe for lower height states.

jsolman replied:
It makes sense. I agree it is fine to leave it this way.

erikzhang commented on neo/Settings.cs:
```diff
@@ -14,6 +14,8 @@ internal class Settings
         public IReadOnlyDictionary<TransactionType, Fixed8> SystemFee { get; private set; }
         public Fixed8 LowPriorityThreshold { get; private set; }
+        public int MaxTransactionsPerBlock { get; private set; }
+        public int MaxFreeTransactionsPerBlock { get; private set; }
```
These two parameters and policy settings may be different, it is best to have a way to reuse.

jsolman replied:
Since the core needs to know these, maybe the plugin policy should read these from the the core.

erikzhang replied:
I suddenly thought that maybe only the consensus nodes need to re-verify all transactions, and there is no need for the normal nodes to re-verify the transactions.

jsolman replied:
I’m up for discussing ideas about the strategy for rebroadcasting. My initial thought is that it should use a random amount of time since last rebroadcast as a threshold to determine whether to rebroadcast again for a transaction; also it should only consider rebroadcasting transactions that have a chance to make it in the next 2 blocks or so based on their sort order in the pool.

vncoelho replied:
@jsolman, I think it is a good idea to limit the re-broadcast for only covering the potential tx for the next 2-3 blocks:
https://github.com/neo-project/neo/blob/831003582b2ada4aff49244e5c88679b8982cc47/neo/Ledger/Blockchain.cs#L382-L398

vncoelho replied:
But, for that, we would need to know the order the possible order of a given  `tx` at `OnPersistCompleted`.
I think that it do not really need to be the exact order, but only to avoid sending some tx would be nice.

vncoelho replied:
Following the same reasoning, I think that honest Seed Nodes should not really send transactions that will surely not enter in the next block 3-5 blocks.

erikzhang replied:
What I mean is that only the consensus nodes need to re-verify `MaxTransactionsPerBlock` number of transactions immediately after the block is persisted. So this can be triggered by the SimplePolicy plugin, so you don't need to write these parameters to the core configuration file.

jsolman replied:
@erikzhang 
Ok, I’ll look into adjusting to receive the params from the plugin. 

It would be nice if there were a way to inspect the maximum number of transactions each of the consensus nodes are configured to verify when they create a block. For now, perhaps the mechanism for determining whether to re-relay the reverified transactions will have to not consider how many transactions can be validated per block. I’ll come up with something that should ensure they don’t get orphaned and also doesn’t bog down the network if the mempool stays full for a long time.

jsolman replied:
@erikzhang I've implemented reading the params from the plugin.

jsolman commented:
I added rebroadcasting transactions upon re-verification if it has passed a certain number of block times since the last rebroadcast. This should meet the needs of ensuring transactions don't get orphaned, while being considerably less taxing on the network than the current code.

erikzhang commented on neo/Ledger/MemoryPool.cs:
```diff
@@ -4,162 +4,504 @@
+
+            ReverifyTransactions(_sortedHighPrioTransactions, _unverifiedSortedHighPriorityTransactions,
+                _maxTxPerBlock, MaxSecondsToReverifyHighPrioTx, snapshot);
+            ReverifyTransactions(_sortedLowPrioTransactions, _unverifiedSortedLowPriorityTransactions,
```
If the node does not have any policy plugins installed, then it will get `int.MaxValue`, which will have to reverify all transactions after block persistence.

In fact, for non-consensus nodes, they do not install policy plugins, and there is no need to reverify transactions after block persistence. In this case, the reverification of the transaction can be handled in the `Idea` message.

jsolman replied:
I already accounted for this, see line 494:
```
           if (Plugin.Policies.Count == 0)
                return;
```

jsolman commented:
I've got this running in a private net now I just need to start bombarding it with transactions.

jsolman commented:
There is one minor change needed in neo-cli to work with this PR. I can put out the PR for that as well if you want, but since I haven't yet, if you want to run it, it is just this one line change:
```
diff --git a/neo-cli/Shell/MainService.cs b/neo-cli/Shell/MainService.cs
index fe93dc8..4c86c2b 100644
--- a/neo-cli/Shell/MainService.cs
+++ b/neo-cli/Shell/MainService.cs
@@ -1008,7 +1008,7 @@ namespace Neo.Shell
         private bool OnShowPoolCommand(string[] args)
         {
             bool verbose = args.Length >= 3 && args[2] == "verbose";
-            Transaction[] transactions = Blockchain.Singleton.GetMemoryPool().ToArray();
+            Transaction[] transactions = Blockchain.Singleton.GetMemoryPoolVerifiedAndSorted().ToArray();
```

jsolman commented:
Making good progress testing.
After each block I have it outputing the counts in the pool
```
 _unsortedTransactions.Count: 20
 _sortedLowPrioTransactions.Count: 20
 _sortedHighPrioTransactions.Count: 0
 _unverifiedTransactions.Count: 6028
 _unverifiedSortedHighPriorityTransactions.Count: 0
 _unverifiedSortedLowPriorityTransactions.Count: 6028
```
It's generating blocks while getting blasted with transactions with a pretty large pool. I'm working my way up to the 50,000 to see how it behaves while full.

jsolman commented:
I'll throw in some high priority ones in there once the pool is full and see that they make it out in 1 block from the time I send them.

jsolman commented:
```
2019-01-10 07:59:39.519 T  Neo.Ledger.Blockchain: Added Transaction to Mempool size=15153
2019-01-10 07:59:39.579 T  Neo.Ledger.Blockchain: Added Transaction to Mempool size=15154
2019-01-10 07:59:39.996 T  Neo.Ledger.Blockchain: OnNewBlock received block Index=3436
2019-01-10 07:59:40.050 D  Neo.Ledger.MemoryPool: After Persist counts:
 _unsortedTransactions.Count: 20
 _sortedLowPrioTransactions.Count: 20
 _sortedHighPrioTransactions.Count: 0
 _unverifiedTransactions.Count: 15116
 _unverifiedSortedHighPriorityTransactions.Count: 0
 _unverifiedSortedLowPriorityTransactions.Count: 15116
2019-01-10 07:59:40.051 T  Neo.Network.P2P.LocalNode: Relaying block index: 3436 to probably 4 remote nodes.
2019-01-10 07:59:40.051 T  Neo.Ledger.Blockchain: Added Transaction to Mempool size=15137
```

jsolman commented:
Consensus happily confirming 20 TX each block
```
[08:45:08.299] initialize: height=3593 view=0 index=3 role=Backup
[08:45:23.301] OnPrepareRequestReceived: height=3593 view=0 index=1 tx=21
[08:45:23.758] OnPrepareResponseReceived: height=3593 view=0 index=2
[08:45:23.846] OnPrepareResponseReceived: height=3593 view=0 index=0
[08:45:25.155] send prepare response
[08:45:25.174] relay block: 0xcbd51f6e4c43b4f03f27901eacd8599aade4189ef8e093316922c62a67e39e84
[08:45:25.389] persist block: 0xcbd51f6e4c43b4f03f27901eacd8599aade4189ef8e093316922c62a67e39e84
[08:45:25.416] initialize: height=3594 view=0 index=3 role=Backup
[08:45:38.636] OnPrepareRequestReceived: height=3594 view=0 index=2 tx=21
[08:45:39.075] OnPrepareResponseReceived: height=3594 view=0 index=0
[08:45:39.364] send prepare response
[08:45:39.393] relay block: 0x258b06536919922494464162426ae53c4ad56d2a27765f03d2b38959b1b6f511
[08:45:39.450] persist block: 0x258b06536919922494464162426ae53c4ad56d2a27765f03d2b38959b1b6f511
[08:45:39.451] initialize: height=3595 view=0 index=3 role=Primary
[08:45:54.462] timeout: height=3595 view=0 state=Primary
[08:45:54.462] send prepare request: height=3595 view=0
[08:45:54.648] OnPrepareResponseReceived: height=3595 view=0 index=2
[08:45:55.183] OnPrepareResponseReceived: height=3595 view=0 index=1
[08:45:55.206] relay block: 0xca33f7457f0db74e8d4ad9a4a8c080cfaa6d49d16137508107d1c197e1ab6f28
[08:45:55.315] persist block: 0xca33f7457f0db74e8d4ad9a4a8c080cfaa6d49d16137508107d1c197e1ab6f28
[08:45:55.317] initialize: height=3596 view=0 index=3 role=Backup
[08:46:11.448] OnPrepareRequestReceived: height=3596 view=0 index=0 tx=21
[08:46:12.290] OnPrepareResponseReceived: height=3596 view=0 index=2
[08:46:25.326] timeout: height=3596 view=0 state=Backup, RequestReceived
[08:46:25.330] request change view: height=3596 view=0 nv=1 state=Backup, RequestReceived, ViewChanging
[08:46:25.333] OnChangeViewReceived: height=3596 view=0 index=2 nv=1
[08:46:25.682] OnChangeViewReceived: height=3596 view=0 index=1 nv=1
[08:46:25.683] changeview: view=1 primary=02103a7f7dd016558597f7960d27c516a4394fd968b9e65155eb4b013e4040406e
[08:46:25.684] initialize: height=3596 view=1 index=3 role=Primary
[08:46:26.825] timeout: height=3596 view=1 state=Primary
[08:46:26.828] send prepare request: height=3596 view=1
[08:46:26.898] OnPrepareResponseReceived: height=3596 view=1 index=1
[08:46:28.338] OnPrepareResponseReceived: height=3596 view=1 index=0
[08:46:28.400] relay block: 0xcfd5ab186784dfbc0c8d51e78ae9c66f794093e50403abbe0499edbeb3c0c649
[08:46:29.512] persist block: 0xcfd5ab186784dfbc0c8d51e78ae9c66f794093e50403abbe0499edbeb3c0c649
[08:46:29.513] initialize: height=3597 view=0 index=3 role=Backup
[08:46:42.141] OnPrepareRequestReceived: height=3597 view=0 index=1 tx=21
[08:46:42.226] send prepare response
[08:46:42.847] OnPrepareResponseReceived: height=3597 view=0 index=0
[08:46:42.850] relay block: 0x17e9570788d11dd3e6c7b330573fed637bd9a30eae3ce77991ad8d5423ceecfc
[08:46:42.916] persist block: 0x17e9570788d11dd3e6c7b330573fed637bd9a30eae3ce77991ad8d5423ceecfc
[08:46:42.917] initialize: height=3598 view=0 index=3 role=Backup
[08:46:59.047] OnPrepareRequestReceived: height=3598 view=0 index=2 tx=21
[08:46:59.134] send prepare response
[08:47:00.387] OnPrepareResponseReceived: height=3598 view=0 index=1
[08:47:00.390] relay block: 0x0476c71cdc7a05a572a7fac6a324af5c3bf8985d8c6e290e03f053d08ed898fa
[08:47:00.866] persist block: 0x0476c71cdc7a05a572a7fac6a324af5c3bf8985d8c6e290e03f053d08ed898fa
[08:47:00.871] initialize: height=3599 view=0 index=3 role=Primary
[08:47:15.877] timeout: height=3599 view=0 state=Primary
[08:47:15.878] send prepare request: height=3599 view=0
[08:47:15.923] OnPrepareResponseReceived: height=3599 view=0 index=0
[08:47:18.966] OnPrepareResponseReceived: height=3599 view=0 index=2
[08:47:18.971] relay block: 0x520d4867dece0692eaf84805cbd113917ad4533d740d357fcb7b12054541d3a6
[08:47:19.152] persist block: 0x520d4867dece0692eaf84805cbd113917ad4533d740d357fcb7b12054541d3a6
```
Pool up to 26K transactions

shargon commented on neo/Network/RPC/RpcServer.cs:
```diff
@@ -297,7 +297,21 @@ private JObject Process(string method, JArray _params)
                     }
                 case "getrawmempool":
-                    return new JArray(Blockchain.Singleton.GetMemoryPool().Select(p => (JObject)p.Hash.ToString()));
+                    {
```
This change affect API implementations, maybe we should change the return logic with a `format` parameter or something like that

jsolman replied:
The format is the exact same if you don't pass the new parameter. I don't think any change is needed from how it is now.

jsolman replied:
See my comment on line 303; how does that not return the same thing after this change as was returned before this change?

jsolman commented:
I hit 50K transactions and found a bug when it removes over-capacity, and just pushed a fix. One more test for over-capacity and then I think this is ready to go.

jsolman commented on neo/Network/RPC/RpcServer.cs:
```diff
@@ -297,7 +297,21 @@ private JObject Process(string method, JArray _params)
+                    {
+                        bool shouldGetUnverified = _params.Count >= 1 && _params[0].AsBooleanOrDefault(false);
+                        if (!shouldGetUnverified)
+                            return new JArray(Blockchain.Singleton.GetMemoryPoolVerified().Select(p => (JObject)p.Hash.ToString()));
```
@shargon this line Iine returns the same thing for `getrawmempool` as before this change; so you should receivie the same thing for the call before or after this change.

jsolman replied:
Only if the new parameter is present will the response format be different from before; therefore, existing clients will not be broken, and new versions of the client that pass with a parameter of true can know how to handle the new response format. So the parameter passed is already like something to specify the format desired.

vncoelho replied:
So there is an option on the setup?
It is an interesting idea, Jeff, because avoid attacks with this type of request.

shargon replied:
You are right xD, sorry for my mistake

vncoelho commented on neo/Ledger/Blockchain.cs:
```diff
@@ -218,11 +217,32 @@ public static UInt160 GetConsensusAddress(ECPoint[] validators)
             return mem_pool;
         }
 
+        public void GetMemoryPoolSeparateVerfiedAndUnverified(out IEnumerable<Transaction> verifiedTransactions,
```
Maybe `GetMemoryPoolSeparateVerfiedAndUnverified` to GetMemoryPoolVerfiedAndUnverifiedPair or GetMemoryPoolPairVerfiedAndUnverified

vncoelho commented on neo/Network/P2P/ProtocolHandler.cs:
```diff
@@ -264,7 +264,7 @@ private void OnInvMessageReceived(InvPayload payload)
         private void OnMemPoolMessageReceived()
         {
-            foreach (InvPayload payload in InvPayload.CreateGroup(InventoryType.TX, Blockchain.Singleton.GetMemoryPool().Select(p => p.Hash).ToArray()))
+            foreach (InvPayload payload in InvPayload.CreateGroup(InventoryType.TX, Blockchain.Singleton.GetMemoryPoolVerified().Select(p => p.Hash).ToArray()))
```
This sounds reasonable, just share verified TX! :+1:

vncoelho commented on neo/Ledger/MemoryPool.cs:
```diff
@@ -4,162 +4,521 @@
-        private readonly ConcurrentDictionary<UInt256, PoolItem> _mem_pool_fee = new ConcurrentDictionary<UInt256, PoolItem>();
-        private readonly ConcurrentDictionary<UInt256, PoolItem> _mem_pool_free = new ConcurrentDictionary<UInt256, PoolItem>();
+        // Allow reverified transactions to be rebroadcast if it has been this many block times since last broadcast.
+        private const int BlocksTillRebroadcastLowPriorityPoolTx = 30;
```
Sounds interesting.

vncoelho commented on neo/Ledger/MemoryPool.cs:
```diff
@@ -4,162 +4,521 @@
+                // NOTE: This really shouldn't be necessary any more and can actually be counter-productive, since
+                //       they can be re-added anyway and would incur a potential addtional verification; setting to 1000
+                //       blocks for now. If transactions want to only be valid for short times they can include a script.
+                var lowPriorityCutOffTime = DateTime.UtcNow.AddSeconds(-Blockchain.SecondsPerBlock * 1000);
```
In this cutoff new in this PR, @jsolman ?

jsolman replied:
There was a cutoff before, but it was broken because it restamped times of transactions when it rebroadcast after blocks. This fixes it.

jsolman replied:
Hmmm. It is an interesting question. I think it really is good if clients include something to expire transactions explicitly after so many blocks. This cutoff time won't really get the transaction out of the network, because other nodes will still re-relay it after this removes it from this node. So it does seem rather pointless to have it.  Since we can handle the pool full now without issues with these changes it also isn't necessary for stability or anything, so I will go ahead and remove it.

vncoelho commented on neo/Ledger/MemoryPool.cs:
```diff
@@ -4,162 +4,521 @@
+            }
+
+            // If we know about headers of future blocks, no point in verifying transactions from the unverified tx pool
+            // until we get caught up.
```
I already said, but, again, this is surely a great improvements that comes with this PR! Congratulations

jsolman commented:
I switched to a TX limit of 5000 now and ran again in a PrivateNet

```
2019-01-10 20:15:54.414 T  Neo.Ledger.Blockchain: Added Transaction to Mempool size=5000
2019-01-10 20:15:54.458 T  Neo.Ledger.Blockchain: Added Transaction to Mempool size=5000
2019-01-10 20:15:54.497 T  Neo.Ledger.Blockchain: Added Transaction to Mempool size=5000
2019-01-10 20:15:54.552 T  Neo.Ledger.Blockchain: Added Transaction to Mempool size=5000
2019-01-10 20:15:54.596 T  Neo.Network.P2P.LocalNode: Telling blockchain about relayed block w/ index: 353.
2019-01-10 20:15:54.599 T  Neo.Ledger.Blockchain: OnNewBlock received block Index=353
2019-01-10 20:15:54.631 D  Neo.Ledger.MemoryPool: After Persist counts:
 _unsortedTransactions.Count: 20
 _sortedLowPrioTransactions.Count: 20
 _sortedHighPrioTransactions.Count: 0
 _unverifiedTransactions.Count: 4960
 _unverifiedSortedHighPriorityTransactions.Count: 0
 _unverifiedSortedLowPriorityTransactions.Count: 4960
2019-01-10 20:15:54.632 T  Neo.Network.P2P.LocalNode: Relaying block index: 353 to probably 5 remote nodes.
2019-01-10 20:15:54.642 T  Neo.Ledger.Blockchain: Added Transaction to Mempool size=4981
2019-01-10 20:15:54.652 T  Neo.Ledger.Blockchain: Added Transaction to Mempool size=4982
2019-01-10 20:15:54.660 T  Neo.Ledger.Blockchain: Added Transaction to Mempool size=4983
```

When the pool fills it now correctly handles discarding the transactions above the limit.

jsolman commented:
Now testing filling the pool with high priority transactions

vncoelho commented:
Great job and report, my friend. Congratulations.

jsolman commented on neo/Ledger/Blockchain.cs:
```diff
@@ -114,14 +111,15 @@ public class ImportCompleted { }
         private readonly Dictionary<uint, LinkedList<Block>> block_cache_unverified = new Dictionary<uint, LinkedList<Block>>();
-        private readonly MemoryPool mem_pool = new MemoryPool(50_000);
-        private readonly ConcurrentDictionary<UInt256, Transaction> mem_pool_unverified = new ConcurrentDictionary<UInt256, Transaction>();
+        private readonly MemoryPool mem_pool;
```
I do not like exposing it because too many methods of the class are not safe to call outside of Blockchain actor.

jsolman replied:
I can create another class that has the methods and expose that if you want.

jsolman replied:
Yes, I’ll do that, but I’m still not sure we want to expose them throughout the project.

jsolman replied:
Actually I already have all the unsafe ones marked `internal`. I guess I was just thinking protecting the rest of the project from calling them outside of `Blockchain`.

jsolman replied:
Done.

shargon commented on neo/Ledger/MemoryPool.cs:
```diff
@@ -1,165 +1,516 @@
+                _txRwLock.EnterReadLock();
+                try
+                {
+                    return _unsortedTransactions.Count + _unverifiedTransactions.Count;
```
We need this lock?

jsolman replied:
Yes.

ThomasLobker commented:
Thanks Jeff, this is great work! It will provide a lot of improved stability.

jsolman commented on neo/Ledger/MemoryPool.cs:
```diff
@@ -1,165 +1,513 @@
 
-        private readonly ConcurrentDictionary<UInt256, PoolItem> _mem_pool_fee = new ConcurrentDictionary<UInt256, PoolItem>();
-        private readonly ConcurrentDictionary<UInt256, PoolItem> _mem_pool_free = new ConcurrentDictionary<UInt256, PoolItem>();
+        // Allow reverified transactions to be rebroadcast if it has been this many block times since last broadcast.
```
I’m going to switch this to use a linear formula that increases if their are more transactions in the pool since rebroadcasting has more network load when the pool is more full.

jsolman replied:
I’ll do it in a separate PR though. I think it is important to get this change merged in now.
